runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_dotprod.c

```diff
@@ -75,3 +75,129 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
     iree_uk_mmt4d_tile_s8s8s32_2x8x4_arm_64_dotprod,
     iree_uk_mmt4d_tile_s8s8s32_4x8x4_arm_64_dotprod,
     iree_uk_mmt4d_tile_s8s8s32_8x8x4_arm_64_dotprod)
+
+static inline void iree_uk_mmt4d_tile_s8s4s32_1x4x8_to_8x4x8_arm_64_dotprod(
```

Comment:
stupid question about API naming
is the x4 in sizes 1x4x8  the KR?
for neon sdot, KR is naturally 4, so is 1x4x8 MR=1 KR=4 NR=8 ?
s8s4s32 means int8 activations, int4 weights, int32 output

Reply:
> is the x4 in sizes 1x4x8 the KR?

No, the convention here is MxNxK.  The 4 here is the N.

Actually I hadn't paid attention. @mariecwhite please explain why the N=4 here, why not N=8?

Reply:
Yeah,

```c
int32x4_t acc[8];
```

So this is currently using only 8 registers for accumulators, out of 32 available. Definitely room to double that to 16. This will keep this kernel closer to other related kernels, will not be much more verbose if you use a `for` loop to un-unroll this code as per my other comment, and will further increase data parallelism, which might improve performance on some current or future CPU but is generally a good thing to have.

Reply:
It will make a particularly big impact in the `M0==1` case. See how the current benchmark result for that case looks anomalously low:
```
BM_mmt4d_s8s4s32_tile_1x4x8_dotprod/real_time         0.614 us        0.613 us      1048575 items_per_second=26.674G/s
```
At least this case will be helped by doubling `N0` from 4 to 8.

Reply:
IIUC (from looking at debug statements), N0 is 4 to match 4 contiguous bytes required for the LHS of the dotprod instruction. K0 is 8 since the RHS includes int4 weights and these are interleaved. In order to get 4 contiguous bytes from the weights, we need to load 8 int4's, de-interleave to get 2 lots of contiguous 4-byte int4 weights.

I briefly tried to use N0=8 instead of N0=4 but this involved unrolling on the outer loop. I'll revisit and see if I can make it work. The `1x4x8` latency is slower than `8x4x8` :(

Reply:
Updated to use `1x8x8`, `2x8x8`, and `4x8x8`. `8x8x8` was taking too many registers. Here are the benchmarks:

Before:
```
-----------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------------------
BM_mmt4d_s8s4s32_tile_1x4x8_dotprod/real_time         0.614 us        0.613 us      1048575 items_per_second=26.674G/s
BM_mmt4d_s8s4s32_tile_2x4x8_dotprod/real_time         0.265 us        0.265 us      4194303 items_per_second=123.484G/s
BM_mmt4d_s8s4s32_tile_4x4x8_dotprod/real_time         0.343 us        0.343 us      2097151 items_per_second=190.916G/s
BM_mmt4d_s8s4s32_tile_8x4x8_dotprod/real_time         0.551 us        0.550 us      1048575 items_per_second=237.73G/s
```

After:
```
-----------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------------------
BM_mmt4d_s8s4s32_tile_1x8x8_dotprod/real_time         0.639 us        0.638 us      1048575 items_per_second=51.2518G/s
BM_mmt4d_s8s4s32_tile_2x8x8_dotprod/real_time         0.369 us        0.368 us      2097151 items_per_second=177.55G/s
BM_mmt4d_s8s4s32_tile_4x8x8_dotprod/real_time         0.540 us        0.539 us      1048575 items_per_second=242.774G/s
```

Reply:
> IIUC (from looking at debug statements), N0 is 4 to match 4 contiguous bytes required for the LHS of the dotprod instruction. K0 is 8 since the RHS includes int4 weights and these are interleaved. In order to get 4 contiguous bytes from the weights, we need to load 8 int4's, de-interleave to get 2 lots of contiguous 4-byte int4 weights.

The LHS tile is M0xK0xi8, the RHS tile is N0xK0xi4 and the accumulator tile is M0xN0xi32, all in row-major layour (the last-enumerated dimension is the inner one, so it's K0 for both LHS and RHS).

So N0 is never the inner dimension of either LHS or RHS. As long as K0==8, both LHS and RHS are guaranteed 4 inner contiguous bytes along the inner accumulation (K) dimension, as needed by dot-product instructions.

> I briefly tried to use N0=8 instead of N0=4 but this involved unrolling on the outer loop. I'll revisit and see if I can make it work.

Going from N0=4 to N0=8 should simply mean 2x-widening what the inner loop body does.

> 8x8x8 was taking too many registers.

Hum. I wonder if that is only a superficial trait of how some variables are declared `int8x8_t` instead of `int8x16_t`, occupying a half register each. Also, it's OK to have some slight over-commitment of registers as long as there exists a register allocation that doesn't involve spilling (ie cycling data registers->stack->registers).

> The 1x4x8 latency is slower than 8x4x8

Of course it is. Narrower tile sizes almost always mean lower performance. But they are the best one can do on narrow problems like `vecmat`. The question is not whether 1x4x8 is slower than 8x4x8, it is whether 1x8x8 is faster than 1x4x8.  Your benchmark shows that it is, by 2x, so the 2x-widening results in a 2x speedup in this case.

> Here are the benchmarks:

Wow, that is a BIG improvement already. Even without the 8x8x8 tile, it is an unconditional improvement, since the 4x8x8 tile goes faster than 8x4x8. You might scrape some additional performance if you find a way to make 8x8x8 work by clever usage of the available registers, usage of `int8x16_t` variables and `vget_low_s8 / vget_high_s8` etc.

Reply:
I added an 8x8x8 which just fits into 30 registers. Here are the latest benchmarks:

```
-----------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------------------
BM_mmt4d_s8s4s32_tile_1x8x8_dotprod/real_time         0.638 us        0.637 us      1048575 items_per_second=51.3324G/s
BM_mmt4d_s8s4s32_tile_2x8x8_dotprod/real_time         0.370 us        0.369 us      2097151 items_per_second=177.164G/s
BM_mmt4d_s8s4s32_tile_4x8x8_dotprod/real_time         0.536 us        0.535 us      1048575 items_per_second=244.61G/s
BM_mmt4d_s8s4s32_tile_8x8x8_dotprod/real_time         0.933 us        0.932 us      1048575 items_per_second=280.844G/s
```

Reply:
> The LHS tile is M0xK0xi8, the RHS tile is N0xK0xi4 and the accumulator tile is M0xN0xi32, all in row-major layout (the last-enumerated dimension is the inner one, so it's K0 for both LHS and RHS. And notice that the "t" in "mmt4d" means "RHS is transposed", which is why K0 is the inner dimension for it as well as for LHS).
So N0 is never the inner dimension of either LHS or RHS. As long as K0==8, both LHS and RHS are guaranteed 4 inner contiguous bytes along the inner accumulation (K) dimension, as needed by dot-product instructions.

Ah this makes a lot of sense now, thanks for the explanation. Both M0 and N0 can scale up without affecting the inner dimension K0.